### PR TITLE
Introduction of CoreScene

### PR DESCRIPTION
**Pull request by deltakosh**
Ok folks, bear with me!

### Introduction

This is the inital commit and before going further I want to be sure we are all on the same page.

This PR will introduce the `CoreScene`, a class that will do the barely minimum work to ensure framework coherence. The goal is to have a super slim class where the user will be ultimately responsible for managing objects lifetime and rendering.

The goal is to allow this kind of code:
```
    // This creates a basic Babylon Scene object (non-mesh)
    var scene = new BABYLON.CoreScene(engine);

    // This creates and positions a free camera (non-mesh)
    var camera = new BABYLON.FreeCamera("camera1", new BABYLON.Vector3(0, 5, -10), scene);

    // This targets the camera to scene origin
    camera.setTarget(BABYLON.Vector3.Zero());

    // This attaches the camera to the canvas
    camera.attachControl(canvas, true);

    // This creates a light, aiming 0,1,0 - to the sky (non-mesh)
    var light = new BABYLON.HemisphericLight("light", new BABYLON.Vector3(0, 1, 0), scene);

    // Default intensity is 1. Let's dim the light a small amount
    light.intensity = 0.7;

    // Our built-in 'sphere' shape.
    var sphere = BABYLON.MeshBuilder.CreateSphere("sphere", {diameter: 2, segments: 32}, scene);

    // Move the sphere upward 1/2 its height
    sphere.position.y = 1;

    // Our built-in 'ground' shape.
    var ground = BABYLON.MeshBuilder.CreateGround("ground", {width: 6, height: 6}, scene);

    scene.render = () => {
        engine.clear(new BABYLON.Color4(0, 0, 0, 1));
        ground.computeWorldMatrix();
        sphere.computeWorldMatrix();
        ground.render();
        sphere.render();
    }
```

As you can see, the user will have to deal with the complete rendering phase.
Not mentioned here but user will also have to listen to scene.onDisposeObservable to dispose all entities (Because `CoreScene` will not have the code to track them. Again because we want a lean class).

### Reasoning

Here we are trading simplicity for both perf, control and size. The size improvement is expected to be quite significant.

On perf side, we also expect it to be quite an improvement if you have simple scenes or simply scenes that do not need all the features we packed in the `Scene` during the years. Worth noting: we will later introduce a new feature that will fit well with the `CoreScene`: the Render Graph where the user will have the freedom to use a render graph instead of manually calling all rendering steps.

### Caveats

The obvious caveats are the following:

- A major breaking change is introduced as all the `***.getScene()` will now return a `CoreScene` instead of a `Scene`. The user can decide to then cast the returned value to `Scene` to mitigate
- Several observables and functions will not be guaranteed anymore and if the user uses `CoreScene`, they will appear as potentially undefined. This one is less of a burden but still an important one as again, the user can cast the `getScene` to `Scene` (because `Scene` remains unchanged with its overrides).
- I need to spend A LOT OF TIME moving everything

### Notes

- JavaScript users should not be impacted as we are only talking about breaking changes at types level
- AbstractScene will be removed from the hierarchy and replace by an interface that `Scene` and `AssetContainer` will implement 
- `CoreScene` can be extended with features like animation support by importing the right file (see here with animatable). So user can choose to start from the `CoreScene` and then extend it based on needs
- `CoreScene` will not replace `Scene`. It is only a new option for users to have more control over the rendering loop
- Lights are not automatically processed so we will have to provide a way to set the active lights on the core scene
- Several use cases will simply not work with `CoreScene` hence why it is flagged as @experimental. The plan is to reach a state where we can compile and merge the PR. We will then decide which features we will want to make compatible with `CoreScene` and how

### Next steps
- Current PR will focus on updating all classes that reference `Scene` and change the reference to `CoreScene`. A walk in the park...
- A new PR will extract some features from `Scene` so users can also add them to `CoreScene` (like input management for instance)

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Animations/animatable.ts [hunk 1]
```diff
@@ -482,11 +484,39 @@ export class Animatable {
     }
 }
 
-declare module "../scene" {
-    export interface Scene {
+declare module "../coreScene" {
+    export interface CoreScene {
```
Comment (RaananW) [inline packages/dev/core/src/Animations/animatable.ts:488]: Just a thought, to fit the basic concept of the core scene (which should have the minimum functionality):
If we are already changing it, an esm-friendly version can be supported as well (abstracted here):

1) Define all of the functions outside of a class
2) Keep the functions defined in Scene, while users can use them outside of the scene when using the corescene.

For example (deliberately taking a public function):

```javascript
// animatable.functions.ts

export function stopAnimation(coreScene: CoreScene, target: any, animationName?: string, targetMask?: (target: any) => boolean): void {
    // MAGIC!
}

// animatable, technically stay as is except for:
import { stopAnimation } from "./animatable.functions.js";
Scene.prototype.stopAnimation = function (target: any, animationName?: string, targetMask?: (target: any) => boolean): void { 
    stopAnimation(this, target .....);
}
```

Another possibility is to not implement these helper functions at all in CoreScene. playAnimation and stopAnimation are nothing more than a helper functions. it is enough to have a reference to the animatable to stop the animation.

**Review by RaananW — COMMENTED** [review]
I think my main question is - what moves to CoreScene and what stays scene-only.

The Scene has a lot of helper functions that are ease-of-use functions. These functions can technically be esm-based functions and don't have to stay in the CoreScene, making the same modifications we are doing to the scene in order to get them to work.

It will be great to remove the side effects when using things like animatables - the functions can be implemented directly in scene (avoiding side effects in animatable.ts) and completely missing from CoreScene, but available outside of the CoreScene as independent helper functions.

I hope this makes sense?

**Comment by deltakosh** [comment]
Unless you meant: keep the current animatable code for Scene but extract the functions so CoreScene can also have them?

**Comment by deltakosh** [comment]
Ok third time the charm:) so if they are in scene, they are no more tree shakable when using Scene. Is it a big concern ?

**Comment by RaananW** [comment]
> Ok third time the charm:) so if they are in scene, they are no more tree shakable when using Scene. Is it a big concern ?

This is what we discussed back then, and exactly what we are doing with "Engine" and its extensions. But we can revisit that for sure :-)

**Comment by deltakosh** [comment]
No no I like it but the problem is for treeshaking then when using Scene. @sebavan thoughts?

**Comment by deltakosh** [comment]
We can certainly consider that animations are really core so maybe everyone wants them for Scene

**Comment by deltakosh** [comment]
But then look at Ray.ts. We certainly do not want it by default in scene so we need to stick with the current import on demand + side effects right?

**Comment by deltakosh** [comment]
Yes but this another big one: it won't compile as I'm replacing scene by CoreScene everywhere and TS will expect these functions to be on CoreScene

**Comment by RaananW** [comment]
We should avoid prototype extensions on `Scene` in classes that might be used by users of the `CoreScene` class.
It's either extending CoreScene's prototype or moving the functions to scene directly, making scene a big phat class. Not sure I like either one of the options.

If we are fine with a slight back-compat change to es6 users we could add the side-effects (the Scene prototype extension) in a separate file (while including it in UMD thus not breaking it). This way you would need to import `Ray.sideEffects.js` to extend the scene's prototype.

It could actually be a positive experience for bundler-users because this way we can automatically define the `sideEffects` in package.json to be `all files with `.sideEffects.` in their name.

**Comment by RaananW** [comment]
> Yes but there is another big one: it won't compile as I'm replacing scene by CoreScene everywhere and TS will expect these functions to be on CoreScene (which is fine as I extend CoreScene prototype for now)

Internally (in the framework itself) we should not use `scene.pick`, we should use the external `pick` function. We don't have to maintain back-compat internally, no?

**Comment by deltakosh** [comment]
No we don't but then do we want to make a super fat Scene class? I doubt it

**Comment by deltakosh** [comment]
> We should avoid prototype extensions on `Scene` in classes that might be used by users of the `CoreScene` class.
> 
> It's either extending CoreScene's prototype or moving the functions to scene directly, making scene a big phat class. Not sure I like either one of the options.
> 
> 
> 
> If we are fine with a slight back-compat change to es6 users we could add the side-effects (the Scene prototype extension) in a separate file (while including it in UMD thus not breaking it). This way you would need to import `Ray.sideEffects.js` to extend the scene's prototype.
> 
> 
> 
> It could actually be a positive experience for bundler-users because this way we can automatically define the `sideEffects` in package.json to be `all files with `.sideEffects.` in their name.

Let's chat offline to see how to do that

**Comment by RaananW** [comment]
> No we don't but then do we want to make a super fat Scene class? I doubt it

nope, this is why i suggested the slight back-compat change to scene. Otherwise we are going to make an unnecessarily large CoreScene function, and not get rid of side effects and prototype extension

**Comment by noname0310** [comment]
I apologize in advance if it doesn't make sense in the flow of the conversation.

Instead of using prototype extensions, you can create scenes with a list of class constructors with different features, which is a really effective way of treeshaking in ES6.

Here's some PlayCanvas code that uses the method I'm talking about
image

This is because this way, users will naturally import the features they need from the engine initialization code.

I think this is the right direction for ES6.

**Comment by ryantrem** [comment]
So if I'm understanding correctly from your comments above:

> ...The user can decide to then cast the returned value to Scene to mitigate
Several observables and functions will not be guaranteed anymore and if the user uses CoreScene, they will appear as potentially undefined...

They can literally cast (e.g. `const scene = someCoreSceneInstance as Scene`), and contractually everything is there, but the actual values may be unexpectedly undefined, yes? If that is the case, I'd be worried that this is too complex and confusing for users. I think either having a core layer throughout the entire API, or having a way to actually "promote" a CoreScene instance to a Scene instance would be needed. I have some ideas on this, but I want to make sure I'm understanding the current state of things in this PR before I go further...

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by deltakosh** [comment]
@ryantrem not really. The user will create the scene. If they create it as a Scene then getScene() as Scene is the only change they have to do. We do not create the scenes. If they create a CoreScene the getScene() will be fine but all the time the users are in control as they know what they created

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by deltakosh** [comment]
Closing that PR as after discussing with the team we decided that the cost on perf, maintenance and back compat will not justify the size gain

**Pull request closed without merging**